src/csharp/Grpc.Core.Tests/CallCancellationTest.cs

```diff
@@ -179,6 +179,122 @@ public async Task ServerStreamingCall_CancelClientSideRead()
             }
         }
 
+        [Test]
+        public async Task ServerStreamingCall_DisposeCallBeforeAllRead()
+        {
+            // Test to reproduce https://github.com/grpc/grpc/issues/8451
+            // "Client channel socket leaks unless read stream drained explicitly"
+            // This version of the test disposes the call before the response stream is
+            // fully read.
+
+            helper.ServerStreamingHandler = new ServerStreamingServerMethod<string, string>(async (request, responseStream, context) => {
+                try
+                {
+                    // send responses to the client with a short delay between each response
+                    foreach (string resp in request.Split(new[] { ' ' }))
+                    {
+                        await responseStream.WriteAsync(resp);
+                        await Task.Delay(50);
+                    }
+                    context.ResponseTrailers.Add("xyz", "");
+                }
+                catch (RpcException ex)
+                {
+                    // server may get an exception when the client cancels
+                    Assert.AreEqual(StatusCode.Cancelled, ex.Status.StatusCode);
+                }
+            });
+
+            using (var call = Calls.AsyncServerStreamingCall(helper.CreateServerStreamingCall(), "A B C D E"))
+            {
+                Assert.AreEqual(1, channel.GetCallReferenceCount());
+
+                try
+                {
+                    while (await call.ResponseStream.MoveNext())
```

Comment:
nit: the while() { do_sth_once; break;) is equivalent to  just invoking call.ResponseStream.MoveNext().
I think just calling MoveNext once without a loop and adding an explanatory comment makes the test shorter and the test logic easier to read.
I don't feel strongly about this though so leaving the resolution up to you.